console/src/components/Services/RemoteSchema/Permissions/PermissionEditor.tsx

```diff
@@ -0,0 +1,369 @@
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import * as GQL from 'graphql';
+import _ from 'lodash';
+import { generateSDL, generateConstantTypes, getChildArgument } from './utils';
+import Button from '../../../Common/Button/Button';
+import styles from '../../../Common/Permissions/PermissionStyles.scss';
+import { DatasourceObject, FieldType } from './types';
+import RSPInput from './RSPInput';
+
+const RootContext = createContext({});
+interface RSPTreeComponentProps {
+  list: FieldType[];
+  setState: (d: FieldType[]) => void;
+}
+type ExpandedItems = {
+  [key: string]: boolean;
+};
+
+const Tree: React.FC<RSPTreeComponentProps> = ({ list, setState }) => {
+  // TODO add checkbox
+  // TODO create and sync tree
+  // TODO check actual gql schema structure and change, if required
+  const [expandedItems, setExpandedItems] = useState<ExpandedItems>({});
+  const onCheck = useCallback(
+    ix => (e: React.FormEvent<HTMLInputElement>) => {
+      const newList = [...list] as FieldType[];
+      newList[ix] = { ...list[ix], checked: e.target.checked };
+      setState([...newList]);
```

Comment:
Let's simplify this!
```suggestion
      setState(prev => prev.map(
        (item, index) => index === ix ? { ...item, checked: e.target.checked } : item
      );
```